### PR TITLE
[WPF] Fix assembly version conflicts

### LINKED ISSUE #10489: [Bug] WPF runtime issue
### Description

FileNotFoundException: Could not load file or assembly 'Xamarin.Forms.Platform.WPF, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.

### Steps to Reproduce

1. Create netcoreapp5.0 app or netcoreapp3.1
2. restore nuget packages (Xamarin.Forms, Xamarin.Forms.Platform.WPF)
3. Create MainPage OnStart
```
var page = new ContentPage();
            var stack = new StackLayout();
            stack.Children.Add(new Label() { Text = "12" });
            page.Content = stack;
            MainPage = new NavigationPage(page);
```
3. Maximize window

### Expected Behavior
app runs

### Actual Behavior
content is not rendered
crash
```
   at System.Windows.FrameworkElement.ChangeLogicalParent(DependencyObject newParent)
   at System.Windows.FrameworkElement.AddLogicalChild(Object child)
   at System.Windows.Controls.UIElementCollection.AddInternal(UIElement element)
   at System.Windows.Controls.UIElementCollection.Add(UIElement element)
   at Xamarin.Forms.Platform.WPF.ScrollViewRenderer.LoadContent()
   at Xamarin.Forms.Platform.WPF.ScrollViewRenderer.OnElementChanged(ElementChangedEventArgs`1 e)
   at Xamarin.Forms.Platform.WPF.ViewRenderer`2.SetElement(VisualElement element)
   at Xamarin.Forms.Platform.WPF.Platform.CreateRenderer(VisualElement element)
   at Xamarin.Forms.Platform.WPF.LayoutRenderer.<>c__DisplayClass4_0.<HandleChildAdded>b__0()
   at Xamarin.Forms.Platform.WPF.Helpers.UiHelper.ExecuteInUiThread(Action action)
   at Xamarin.Forms.Platform.WPF.LayoutRenderer.HandleChildAdded(Object sender, ElementEventArgs e)
   at Xamarin.Forms.Platform.WPF.LayoutRenderer.OnElementChanged(ElementChangedEventArgs`1 e)
   at Xamarin.Forms.Platform.WPF.ViewRenderer`2.SetElement(VisualElement element)
   at Xamarin.Forms.Platform.WPF.Platform.CreateRenderer(VisualElement element)
   at Xamarin.Forms.Platform.WPF.LayoutRenderer.<>c__DisplayClass4_0.<HandleChildAdded>b__0()
   at Xamarin.Forms.Platform.WPF.Helpers.UiHelper.ExecuteInUiThread(Action action)
   at Xamarin.Forms.Platform.WPF.LayoutRenderer.HandleChildAdded(Object sender, ElementEventArgs e)
   at Xamarin.Forms.Platform.WPF.LayoutRenderer.OnElementChanged(ElementChangedEventArgs`1 e)
   at Xamarin.Forms.Platform.WPF.ViewRenderer`2.SetElement(VisualElement element)
   at Xamarin.Forms.Platform.WPF.Platform.CreateRenderer(VisualElement element)
   at Xamarin.Forms.Platform.WPF.Platform.GetOrCreateRenderer(VisualElement element)
   at Xamarin.Forms.Platform.WPF.PageRenderer.UpdateContent()
   at Xamarin.Forms.Platform.WPF.PageRenderer.OnElementChanged(ElementChangedEventArgs`1 e)
   at Xamarin.Forms.Platform.WPF.ViewRenderer`2.SetElement(VisualElement element)
   at Xamarin.Forms.Platform.WPF.Platform.CreateRenderer(VisualElement element)
   at Xamarin.Forms.Platform.WPF.Platform.GetOrCreateRenderer(VisualElement element)
   at Xamarin.Forms.Platform.WPF.FormsContentLoader.CreateOrResizeContent(FrameworkElement parent, VisualElement visualElement)
   at Xamarin.Forms.Platform.WPF.FormsContentLoader.OnSizeContentChanged(FrameworkElement parent, Object page)
   at Xamarin.Forms.Platform.WPF.Controls.FormsContentControl.LightContentControl_SizeChanged(Object sender, SizeChangedEventArgs e)
   at System.Windows.SizeChangedEventArgs.InvokeEventHandler(Delegate genericHandler, Object genericTarget)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEvent(RoutedEventArgs e)
   at System.Windows.FrameworkElement.OnRenderSizeChanged(SizeChangedInfo sizeInfo)
   at System.Windows.ContextLayoutManager.fireSizeChangedEvents()
   at System.Windows.ContextLayoutManager.UpdateLayout()
   at System.Windows.UIElement.UpdateLayout()
   at System.Windows.Interop.HwndSource.Process_WM_SIZE(UIElement rootUIElement, IntPtr hwnd, WindowMessage msg, IntPtr wParam, IntPtr lParam)
   at System.Windows.Interop.HwndSource.LayoutFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.CallWindowProc(IntPtr wndProc, IntPtr hWnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.HwndSubclass.DefWndProcWrapper(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.CallWindowProc(IntPtr wndProc, IntPtr hWnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
```

### Basic Information

- Nuget Packages: 4.8.0-740-nightly

### LINKED ISSUE #11328: [Bug] netcoreapp3.1 WPF version causes Could not load file or assembly 'xxx, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null'
### Description
Any software using netcoreapp3.1/WPF and Xamarin.Forms fails to execute since version 4.7.
Usually the error is Could not load file or assembly 'Xamarin.Forms.Platform.WPF, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null' but it can be 0.0.0.0 depending on the version of the package.
It's not working either with the project 'Xamarin.Forms.ControlGallery.WPF' included in the repository.

### Steps to Reproduce
1. Download any version of the repository since 4.7
2. Build everything
3. Open Visual Studio 2019 with the Xamarin.Forms.sln
4. Set Xamarin.Forms.ControlGallery.WPF as starting project
5. Execute the project

### Expected Behavior
Should run the application

### Actual Behavior
In case of last version 5 nightly, fails with an error "Could not load file or assembly 'Xamarin.Forms.ControlGallery.WPF, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null'".
Any other software made with these packages gives Could not load file or assembly 'Xamarin.Forms.Platform.WPF, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'

### I Couldn't find any workaround and had to downgrade to 4.6.0.1073

### PR DESCRIPTION
**Pull request by krdmllr**
### Description of Change ###

There are currently assembly version missmatches when working with the WPF .NET Core backend. 
The modifications to the assembly version information in this PR solve these issues when running the control gallery application. This is more of a workaround currently and I hope this could help in finding a longterm solution.
Some things I currently dont understand:
* Why is the control gallery explicitly asking for Version 2.0.0.0 of the WPF platform assembly. I could not find a reference to this version anywhere and a runtime error occures when I change the version to something else.
* In the Xamarin.Forms.Platform.WPF project we have to set `<GenerateAssemblyInfo>false</GenerateAssemblyInfo>`, otherwise the following compile error will show up: "The specified version string does not conform to the recommended format - major.minor.build.revision". When setting a version using the Version property, this error persists. In the control gallery, we can and have to set `<GenerateAssemblyInfo>true</GenerateAssemblyInfo>`.
* Why does this error occur in the control gallery? I created an empty wpf .net core and .net framework project, referenced the assemblies and could start up the code shown in https://github.com/xamarin/Xamarin.Forms/issues/10489 without modifications to the assembly infos.

### Issues Resolved ### 
* fixes #11328
* fixes #10489

### Platforms Affected ### 

- WPF

None

Not applicable

### Testing Procedure ###

* Clean bin folder of the Xamarin.Forms.Platform.WPF and Xamarin.Forms.ControlGallery.WPF project
* Build and run
* Application should show up.

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Comment by krdmllr** [comment]
@PureWeen @jsuarezruiz @MagicAndre1981 fyi, you pinged me on this issue. I wont get around working further on this for the next week but perhaps it could help you to further investigate the issue. I suspect this has something to do with the WPF .NET Core tooling and how it creates assembly information, maybe someone from that team could help?

**Comment by jsuarezruiz** [comment]
Tested and Core Gallery is working:

Captura de pantalla 2020-08-25 a las 14 03 05

However, as Konrad said, this changes fix the issue with Core Gallery but we still need to know what is producing the assembly version missmatches.  Thanks for all the effort Konrad, I am not going to approve the PR for now (unless we want this PR now to fix Core Gallery and focus on fix the missmatche issue in another one). I will review the csproj of WPF Platform and Map projects.

**Comment by mattleibow** [comment]
Might be related to this: https://github.com/dotnet/wpf/issues/1688

**Comment by mattleibow** [comment]
The version number comes from here: https://github.com/xamarin/Xamarin.Forms/blob/719fc7a604ff0cce8922d717c99bfb0fa17e35e0/Version.targets#L49

File: Xamarin.Forms.Platform.WPF/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -71,4 +72,6 @@
 
 // Others
 [assembly: Xamarin.Forms.Dependency(typeof(ResourcesProvider))]
-[assembly: Xamarin.Forms.Dependency(typeof(Deserializer))]
\ No newline at end of file
+[assembly: Xamarin.Forms.Dependency(typeof(Deserializer))]
+
+[assembly: AssemblyVersion("2.0.0.0")]
```
Comment (mattleibow) [inline Xamarin.Forms.Platform.WPF/Properties/AssemblyInfo.cs:77]: We shouldn't need both. The `GenerateAssemblyInfo` should automatically include this attribute. I get a compiler error with it...

**Comment by mattleibow** [comment]
@jsuarezruiz  Did you test both netcore and netfx? My netfx still crashes, but netcore is fine...

**Comment by krdmllr** [comment]
> @jsuarezruiz Did you test both netcore and netfx? ~My netfx still crashes, but netcore is fine...~
> 
> My netcoreapp crashes, but netfx is fine.

How does netcore crash? Did you clear bin/obj folders of the platform and galery project?

**Comment by krdmllr** [comment]
> The version number comes from here:
> 
> https://github.com/xamarin/Xamarin.Forms/blob/719fc7a604ff0cce8922d717c99bfb0fa17e35e0/Version.targets#L49

So:
1. Why is the Assembly Version fixed?
2. When is this picked up and why is it not picked up when the wpl platform is build?
3. If the version is fixed, is it ok if we hardcode it in the platform project?

**Comment by PureWeen** [comment]
@rmarinho  @samhouts ?

Do you know why Version.targets isn't being picked up on the WPF dll?

https://github.com/xamarin/Xamarin.Forms/pull/11910#issuecomment-685099324

**Comment by krdmllr** [comment]
I digged deeper into the version file and think that the set version target has some order issue. I added some messages to understand what the $(Version) value is + disabled GenerateAssemblyVersion it works:
image

I then changed the target order to verify that it should only run once after the git info target, which kinda worked but now the dll gets correctly generated!
image

The platform dll now has a usefull file version.
image

And the control gallery runs!

@rmarinho do you know how we could properly change the target order to make this work? I dont want to break other stuff where the beforetargets were required.

**Comment by rmarinho** [comment]
I think this can work, can we keep the BeforeTargets too ?

**Comment by krdmllr** [comment]
> I think this can work, can we keep the BeforeTargets too ?

I could only get it to work without BeforeTargets. Do we need them or could we maybe use other targets?

**Comment by rmarinho** [comment]
@krdmllr can you take a look at #12011

**Comment by krdmllr** [comment]
@rmarinho setting it like in your pr does not build for me

**Comment by rmarinho** [comment]
Hi @krdmllr  so i asked @jsuarezruiz  to test my PR and he was able to run it.. 

Does it fail the same way for you? did you do a git  clean to the repo to make sure old dll is not around?

**Comment by krdmllr** [comment]
@rmarinho @jsuarezruiz yes I cleared everything. Perhaps you did not enable the GenerateAssemblyInfo in the wpf project? We have to enable that but it also creates the error with the current versions file.

**Comment by rmarinho** [comment]
@krdmllr i only pull the branch without any change.

**Comment by rmarinho** [comment]
@krdmllr ok my solution also doesn't seem to be perfect. Doesn't sett the current version ok 

I tested this PR as is, and we generate a 2.0.0.0 fixed assembly and no file version

we can kind of fix it by adding still not the same as we have on targets:
```
[assembly: AssemblyVersion("2.0.0.0")]
[assembly: AssemblyFileVersion(ThisAssembly.Git.BaseVersion.Major + "." 
			+ ThisAssembly.Git.BaseVersion.Minor + "." 
			+ ThisAssembly.Git.BaseVersion.Patch + "." 
			+ ThisAssembly.Git.Commits)]
[assembly: AssemblyInformationalVersion(ThisAssembly.Git.SemVer.Major + "."
			+ ThisAssembly.Git.SemVer.Minor + "."
			+ ThisAssembly.Git.SemVer.Patch
			+ ThisAssembly.Git.SemVer.DashLabel + "+"
			+ ThisAssembly.Git.Commits + "-sha."
			+ ThisAssembly.Git.Commit)]
```

Your solution for the targets file seems to not work for me it says theres a target cycle loop since AfterTargets and DependTargets are same

File: Xamarin.Forms.Platform.WPF/Properties/AssemblyInfo.cs (hunk 1 above)
Comment (rmarinho) [inline Xamarin.Forms.Platform.WPF/Properties/AssemblyInfo.cs:77]:
```suggestion
[assembly: AssemblyVersion("2.0.0.0")]
[assembly: AssemblyFileVersion(ThisAssembly.Git.BaseVersion.Major + "."
			+ ThisAssembly.Git.BaseVersion.Minor + "."
			+ ThisAssembly.Git.BaseVersion.Patch + "."
			+ ThisAssembly.Git.Commits)]
[assembly: AssemblyInformationalVersion(ThisAssembly.Git.SemVer.Major + "."
			+ ThisAssembly.Git.SemVer.Minor + "."
			+ ThisAssembly.Git.SemVer.Patch
			+ ThisAssembly.Git.SemVer.DashLabel + "+"
			+ ThisAssembly.Git.Commits + "-sha."
			+ ThisAssembly.Git.Commit)]
```

**Review by rmarinho — CHANGES_REQUESTED** [review]

**Review by rmarinho — APPROVED** [review]
We are going with this solution for now, but will investigate how to fix to use the Version.targets

**Pull request merged**